yjit/src/core.rs

```diff
@@ -573,24 +573,22 @@ pub extern "C" fn rb_yjit_iseq_update_references(payload: *mut c_void) {
             // Walk over references to objects in generated code.
             for offset in &block.gc_object_offsets {
                 let offset_to_value = offset.as_usize();
-                let value_address: *const u8 = cb.get_ptr(offset_to_value).raw_ptr();
+                let value_code_ptr = cb.get_ptr(offset_to_value);
+                let value_ptr: *const u8 = value_code_ptr.raw_ptr();
                 // Creating an unaligned pointer is well defined unlike in C.
-                let value_address = value_address as *mut VALUE;
+                let value_ptr = value_ptr as *mut VALUE;
 
                 // SAFETY: these point to YJIT's code buffer
-                let object = unsafe { value_address.read_unaligned() };
+                let object = unsafe { value_ptr.read_unaligned() };
                 let new_addr = unsafe { rb_gc_location(object) };
 
-                // Only write when the VALUE moves, to be CoW friendly.
+                // Only write when the VALUE moves, to be copy-on-write friendly.
                 if new_addr != object {
-                    // Possibly unlock the page we need to update
-                    cb.mark_position_writable(offset_to_value);
-
-                    // Object could cross a page boundary, so unlock there as well
-                    cb.mark_position_writable(offset_to_value + size_of::<VALUE>() - 1);
-
-                    // SAFETY: we just made this address writable
-                    unsafe { value_address.write_unaligned(new_addr) };
+                    for (byte_idx, &byte) in new_addr.as_u64().to_le_bytes().iter().enumerate() {
+                        let byte_code_ptr = value_code_ptr.add_bytes(byte_idx);
+                        cb.get_mem().write_byte(byte_code_ptr, byte)
+                            .expect("patching existing code should be in-bound");
```

Comment:
Nitpick: should say within bounds because in-bounds is an existing english word.